File: tools/infractl/cmd/docker.go
package cmd

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/spf13/cobra"
)

// NewDockerCommand returns the `docker` subcommand tree.
func NewDockerCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "docker",
		Short: "Manage Docker containers on remote hosts",
		Long:  "Run Docker commands on remote baremetal/VM hosts over SSH.",
	}

	cmd.AddCommand(newDockerLsCommand())
	cmd.AddCommand(newDockerStartCommand())
	cmd.AddCommand(newDockerStopCommand())
	cmd.AddCommand(newDockerRestartCommand())
	cmd.AddCommand(newDockerRmCommand())
	cmd.AddCommand(newDockerLogsCommand())
	cmd.AddCommand(newDockerExecCommand())
	cmd.AddCommand(newDockerPullCommand())
	cmd.AddCommand(newDockerStatsCommand())
	cmd.AddCommand(newDockerRunCommand())

	return cmd
}

func newDockerLsCommand() *cobra.Command {
	var all bool
	return &cobra.Command{
		Use:     "ls <host>",
		Aliases: []string{"ps", "list"},
		Short:   "List containers on a remote host",
		Args:    cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			format := `table {{.ID}}\t{{.Names}}\t{{.Image}}\t{{.Status}}\t{{.Ports}}`
			if all {
				return dockerRun(args[0], "docker ps -a --format '"+format+"'")
			}
			return dockerRun(args[0], "docker ps --format '"+format+"'")
		},
	}
}

// nolint flag defined but not wired — add with PersistentPreRunE if needed
func init() {}

func newDockerStartCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "start <host> <container> [container...]",
		Short: "Start one or more containers on a remote host",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			containers := strings.Join(args[1:], " ")
			return dockerRun(args[0], "docker start "+containers)
		},
	}
}

func newDockerStopCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "stop <host> <container> [container...]",
		Short: "Stop one or more containers on a remote host",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			containers := strings.Join(args[1:], " ")
			return dockerRun(args[0], "docker stop "+containers)
		},
	}
}

func newDockerRestartCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "restart <host> <container> [container...]",
		Short: "Restart one or more containers on a remote host",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			containers := strings.Join(args[1:], " ")
			return dockerRun(args[0], "docker restart "+containers)
		},
	}
}

func newDockerRmCommand() *cobra.Command {
	var force bool
	cmd := &cobra.Command{
		Use:   "rm <host> <container> [container...]",
		Short: "Remove one or more containers on a remote host",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			flag := ""
			if force {
				flag = "-f "
			}
			containers := strings.Join(args[1:], " ")
			return dockerRun(args[0], "docker rm "+flag+containers)
		},
	}
	cmd.Flags().BoolVarP(&force, "force", "f", false, "Force remove running container")
	return cmd
}

func newDockerLogsCommand() *cobra.Command {
	var follow bool
	var tail string
	var since string

	cmd := &cobra.Command{
		Use:   "logs <host> <container>",
		Short: "Fetch logs from a container on a remote host",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			flags := ""
			if follow {
				flags += " -f"
			}
			if tail != "" {
				flags += " --tail " + tail
			}
			if since != "" {
				flags += " --since " + since
			}
			return dockerStream(args[0], "docker logs"+flags+" "+args[1])
		},
	}
	cmd.Flags().BoolVarP(&follow, "follow", "f", false, "Follow log output")
	cmd.Flags().StringVar(&tail, "tail", "100", "Number of lines to show from end")
	cmd.Flags().StringVar(&since, "since", "", "Show logs since timestamp (e.g. 2h, 2024-01-01)")
	return cmd
}

func newDockerExecCommand() *cobra.Command {
	var interactive bool

	cmd := &cobra.Command{
		Use:   "exec <host> <container> [command...]",
		Short: "Run a command in a container on a remote host",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			host := args[0]
			container := args[1]
			execCmd := "sh"
			if len(args) > 2 {
				execCmd = strings.Join(args[2:], " ")
			}
			flag := ""
			if interactive {
				flag = "-it "
			}
			return dockerStream(host, "docker exec "+flag+container+" "+execCmd)
		},
	}
	cmd.Flags().BoolVarP(&interactive, "interactive", "i", false, "Keep stdin open")
	return cmd
}

func newDockerPullCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "pull <host> <image>",
		Short: "Pull a Docker image on a remote host",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			return dockerRun(args[0], "docker pull "+args[1])
		},
	}
}

func newDockerStatsCommand() *cobra.Command {
	var noStream bool
	cmd := &cobra.Command{
		Use:   "stats <host>",
		Short: "Display live resource usage for containers on a remote host",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			flag := ""
			if noStream {
				flag = "--no-stream "
			}
			return dockerStream(args[0], "docker stats "+flag)
		},
	}
	cmd.Flags().BoolVar(&noStream, "no-stream", true, "Disable streaming stats (show one snapshot)")
	return cmd
}

func newDockerRunCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "run <host> -- <docker-run-args...>",
		Short: "Run a docker run command on a remote host",
		Long: `Pass docker run arguments directly.

Example:
  infractl docker run web-01 -- -d --name myapp -p 8080:80 nginx:alpine`,
		DisableFlagParsing: true,
		Args:               cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			host := args[0]
			// Strip '--' separator if present
			rest := args[1:]
			if len(rest) > 0 && rest[0] == "--" {
				rest = rest[1:]
			}
			return dockerRun(host, "docker run "+strings.Join(rest, " "))
		},
	}
}

// dockerRun runs a docker command on a remote host and prints output.
func dockerRun(hostname, cmd string) error {
	ex, err := executorFor(hostname)
	if err != nil {
		return err
	}
	out, err := ex.Run(context.Background(), cmd)
	fmt.Print(out)
	if err != nil && strings.TrimSpace(out) == "" {
		return fmt.Errorf("ssh: %w", err)
	}
	return nil
}

// dockerStream runs a docker command with streaming I/O (for logs -f, exec, stats).
func dockerStream(hostname, cmd string) error {
	ex, err := executorFor(hostname)
	if err != nil {
		return err
	}
	return ex.Stream(context.Background(), cmd, os.Stdout, os.Stderr)
}

File: tools/adsops/src/adsops/infractl/docker.py
"""Docker commands over SSH for infractl.

Provides Python equivalents of the Go infractl docker subcommands,
executing docker CLI commands on remote hosts via asyncssh.

Threat model:
- T-02-05: Container names are passed as separate arguments; validate names
  contain only [a-zA-Z0-9_.-] to prevent shell injection when composing
  command strings. Hosts are validated to contain no shell metacharacters.
"""
import asyncio
import re
import sys

from adsops.infractl.ssh import run_sync, stream_command

# Safe hostname pattern: alphanumeric, dots, hyphens, underscores (no shell metacharacters)
_SAFE_HOST_RE = re.compile(r"^[a-zA-Z0-9._\-]+$")
# Safe container/image name pattern (Docker naming rules + subset of image tag chars)
_SAFE_NAME_RE = re.compile(r"^[a-zA-Z0-9_.\-/:@]+$")


def _validate_host(host: str) -> None:
    """Raise ValueError if host contains shell metacharacters (T-02-05)."""
    if not _SAFE_HOST_RE.match(host):
        raise ValueError(
            f"Invalid hostname {host!r}: must match [a-zA-Z0-9._-]"
        )


def _validate_name(name: str, label: str = "name") -> None:
    """Raise ValueError if a container/image name contains shell metacharacters."""
    if not _SAFE_NAME_RE.match(name):
        raise ValueError(
            f"Invalid {label} {name!r}: must match [a-zA-Z0-9_.\\-/:@]"
        )


_DOCKER_PS_FORMAT = (
    r"table {{.ID}}\t{{.Names}}\t{{.Image}}\t{{.Status}}\t{{.Ports}}"
)


def docker_ls(host: str, all_containers: bool = False) -> tuple[str, str]:
    """List Docker containers on a remote host.

    Args:
        host: Hostname or SSH alias.
        all_containers: If True, pass --all flag to show stopped containers.

    Returns:
        Tuple of (stdout, stderr) from the remote docker ps command.
    """
    _validate_host(host)
    flag = "--all " if all_containers else ""
    cmd = f"docker ps {flag}--format '{_DOCKER_PS_FORMAT}'"
    return run_sync(host, cmd)


def docker_start(host: str, containers: list[str]) -> tuple[str, str]:
    """Start one or more containers on a remote host.

    Args:
        host: Hostname or SSH alias.
        containers: List of container names to start.

    Returns:
        Tuple of (stdout, stderr) from the remote docker start command.
    """
    _validate_host(host)
    for c in containers:
        _validate_name(c, "container name")
    names = " ".join(containers)
    return run_sync(host, f"docker start {names}")


def docker_stop(host: str, containers: list[str]) -> tuple[str, str]:
    """Stop one or more containers on a remote host.

    Args:
        host: Hostname or SSH alias.
        containers: List of container names to stop.

    Returns:
        Tuple of (stdout, stderr) from the remote docker stop command.
    """
    _validate_host(host)
    for c in containers:
        _validate_name(c, "container name")
    names = " ".join(containers)
    return run_sync(host, f"docker stop {names}")


def docker_restart(host: str, containers: list[str]) -> tuple[str, str]:
    """Restart one or more containers on a remote host.

    Args:
        host: Hostname or SSH alias.
        containers: List of container names to restart.

    Returns:
        Tuple of (stdout, stderr) from the remote docker restart command.
    """
    _validate_host(host)
    for c in containers:
        _validate_name(c, "container name")
    names = " ".join(containers)
    return run_sync(host, f"docker restart {names}")


def docker_logs(
    host: str,
    container: str,
    follow: bool = False,
    tail: str = "100",
) -> None:
    """Stream logs from a container on a remote host.

    Args:
        host: Hostname or SSH alias.
        container: Container name or ID.
        follow: If True, follow log output (-f flag).
        tail: Number of lines from end to show (default: 100).
    """
    _validate_host(host)
    _validate_name(container, "container name")
    flags = ""
    if follow:
        flags += " -f"
    if tail:
        flags += f" --tail {tail}"
    cmd = f"docker logs{flags} {container}"
    asyncio.run(stream_command(host, cmd))


def docker_exec(host: str, container: str, cmd: str = "sh") -> None:
    """Execute a command in a running container on a remote host.

    Args:
        host: Hostname or SSH alias.
        container: Container name or ID.
        cmd: Command to run inside the container (default: sh).
    """
    _validate_host(host)
    _validate_name(container, "container name")
    asyncio.run(stream_command(host, f"docker exec {container} {cmd}"))
